listeners/FilterTimeslot.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+namespace Igniter\Local\Listeners;
+
+use Admin\Models\Orders_model;
+use Carbon\Carbon;
+use DateTime;
+use Igniter\Flame\Location\Models\AbstractLocation;
+use Igniter\Local\Classes\Location;
+use Igniter\Local\Facades\Location as LocationFacade;
+use Igniter\Flame\Traits\EventEmitter;
+use Illuminate\Contracts\Events\Dispatcher;
+
+class FilterTimeslot
+{
+    use EventEmitter;
+
+    protected static $ordersCache = [];
+    protected static $scheduleCache = [];
+
+    public function subscribe(Dispatcher $dispatcher)
+    {
+        $dispatcher->listen('igniter.workingSchedule.timeslotFilter', __CLASS__.'@filterTimeslot');
+    }
+
+    public function filterTimeslot($workingSchedule, $timeslot)
+    {
+        // Skip if the working schedule is not for delivery or pickup
+        if ($workingSchedule->getType() == AbstractLocation::OPENING)
+            return;
+            
+        // allow the timeslot selection logic to be overwritten by extensions
+        if ($event = $this->fireSystemEvent('igniter.local.timeslotValid', [$timeslot])) {
```

Comment:
Isn't this offering the same as the `igniter.workingSchedule.timeslotFilter`?

Reply:
If they use `igniter.workingSchedule.timeslotFilter` does it stop this code from running? If so great - that was all I was trying to provide.

Reply:
Listening on `igniter.workingSchedule.timeslotFilter` is what runs the code and `igniter.local.timeslotValid` won't fire unless `igniter.workingSchedule.timeslotFilter` is fired first.

Reply:
Sorry, I'll explain better :)

If someone wanted to add an extension that modified this behaviour of validating timeslots - e.g. say I wanted an extension that handled the 'covers' concept I explained in the roadmap issue, and they listened for the `igniter.workingSchedule.timeslotFilter` event, would that prevent `filterTimeslot()` from running or would both run?

Reply:
That depends since the event dispatcher will stop processing other listeners once one returns a value. eg, if the timeslot passes the order limit check (nothing is returned from filterTimeslot()), it will move to the next listener to check the 'covers'.


Both would run, they'd just have to add a new listener to subscribe to the igniter.workingSchedule.timeslotFilter event. So when the event fires, it loops through all the subscribed listeners and stops on the first listener that returns a value

Reply:
In that case I would be inclined to leave this event to allow the flexibility - but its your call! 😄

Reply:
Yeah that approach would work - thanks for bearing with me!

Reply:
No problem at all!

Reply:
Changing the event to halt when false is returned doesn't seem to do the trick. I thought I could do something like this

```
$result = Event::fire('igniter.workingSchedule.timeslotFilter', [$this, $dateTime]);
return $result === array_filter($result);
```

but that won't work since `false` will never be a value of $result array. 

We can continue with the event halting approach and restrict listeners to only return false when a timeslot is invalid and nothing when valid so the next listener gets fired. Looking at the filterTimeslot listener, you'll need to make some adjustment so it only returns false when order limit has been reached.

I've changed the event name to igniter.workingSchedule.timeslotValid to reflect.

Reply:
No problem - I've pushed an update